launch_testing/launch_testing/pytest/plugin.py

```diff
@@ -0,0 +1,528 @@
+            fixturename = fixture.__name__
+            scope = fixture._pytestfixturefunction.scope
+            is_shutdown = has_shutdown_kwarg(item)
+            items = generate_test_items(collector, name, obj, fixturename, is_shutdown, False)
+            if need_shutdown_test_item(obj) and scope != 'function':
+                # for function scope we only need one shutdown item
+                # if not we're going to use two event loops!!!
+                shutdown_items = generate_test_items(collector, name, obj, fixturename, True, True)
+                for item, shutdown_item in zip(items, shutdown_items):
+                    item._launch_testing_shutdown_item = shutdown_item
+                items.extend(shutdown_items)
+            return items
+
+
+def get_error_context_from_obj(obj):
+    """Return formatted information of the object location."""
+    try:
+        fspath = inspect.getsourcefile(obj)
+    except TypeError:
+        return 'location information of the object not found'
+    try:
+        lines, lineno = inspect.getsourcelines(obj)
+    except IOError:
+        return f'file {fspath}: source code not available'
+    error_msg = f'file {fspath}, line {lineno}'
+    for line in lines:
+        line = line.rstrip()
+        error_msg += f'\n  {line}'
+        if line.lstrip().startswith('def'):
+            break
+    return error_msg
+
+
+def is_shutdown_test(item):
+    """Return `True` if the item is a launch test."""
+    return getattr(item, '_launch_testing_is_shutdown', False)
+
+
+def is_same_launch_test_fixture(left_item, right_item):
+    """Return `True` if both items are using the same fixture with the same parameters."""
+    lfn = get_launch_test_fixture(left_item)
+    rfn = get_launch_test_fixture(right_item)
+    if None in (lfn, rfn):
+        return False
+    if lfn != rfn:
+        return False
+    if lfn._pytestfixturefunction.scope == 'function':
+        return False
+    name = lfn.__name__
+
+    def get_fixture_params(item):
+        if getattr(item, 'callspec', None) is None:
+            return None
+        return item.callspec.params.get(name, None)
+    return get_fixture_params(left_item) == get_fixture_params(right_item)
+
+
+@pytest.mark.trylast
+def pytest_collection_modifyitems(session, config, items):
+    """Move shutdown tests after normal tests."""
+
+    def enumerate_reversed(sequence):
+        # reversed(enumerate(sequence)), doesn't work
+        # here a little generator for that
+        n = len(sequence) - 1
+        for elem in sequence[::-1]:
+            yield n, elem
+            n -= 1
+
+    for i, item in enumerate_reversed(items):
+        # This algo has worst case Nitems * Nlaunchtestitems time complexity.
+        # We could probably find something better, but it's not that easy because:
+        # - Tests using the same launch fixture might not be already grouped,
+        #   i.e. there might be a test not using the fixture in the middle.
```

Comment:
@ivanpauno meta: if this is true, what prevents a module scope launch fixture and a function scope launch fixture from running concurrently?

Reply:
IIUC, it's possible to run into that issue if you have weird dependencies between fixtures.